Guides/Debugging page (#216)
withastro/docs: src/pages/en/guides/debugging.md

description: Debug in Astro using the Debug component
---

> There are a few ways to debug your code with Astro. But it is important to remember that because **Astro runs on the server**, any code that is executed within the `---` Astro Front-matter is `console.log()` to the **terminal**.

Reviewer: “frontmatter” seems to be all one word elsewhere. What do you think about mentioning the browser console explicitly as that’s the other place you might think or hope to see things logged.

```suggestion
There are a few ways to debug your code with Astro. But it is important to remember that because **Astro runs on the server**, any code that is executed within the Astro frontmatter will be logged to the **terminal**, not to the browser console.
```

Reviewer 2: Haha, yeah, it was more fun to type than to read, for sure.

And I agree with both of you @aFuzzyBear  @delucis -- no image (12 reasons, maybe 15...) but since this is unfamiliar territory, I wasn't sure what might be expected in terms of preparing the user for the debugging output.

If I want to get super-technical, is *any* code executed in the front matter logged to the terminal, or just any console.log() output?

Reviewer: 1. Code executed in the frontmatter is just executed, `console.log` is an API for printing — a.k.a. logging — values. In the case of the frontmatter, it passes them to the terminal’s [`stdout`](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)). (In the browser the same API prints to the browser console, which I don’t think has the same concept of `stdout` as terminals.) Actually here’s the `console.log` definition from Node’s docs: “Prints to `stdout` with newline.”

2. I could back that, sure. The formatting does look slightly odd, so maybe just saying that it will render debugging information on the page is enough.

Reviewer 2: So, "any code that is executed within the Astro frontmatter will be logged to the **terminal**" is not *technically* correct... that code is executed, but only stuff explicitly `console.log()`'d  is logged to the terminal, right? That's what I'm trying to verify.  :)

(That would have been my guess, but I do want to make sure here.)

Reviewer: Ohhh, right, yes, you’ve got it spot on.

Reviewer 2: OK, someone please put me out of my misery and make this both accurate, and a sensible sentence: @aFuzzyBear @delucis 

because **Astro runs on the server**, the output of `console.log()` written in Astro frontmatter will be logged to the **terminal**, not to the browser console.

Reviewer: I think this is good! Slight tweak:

> …because Astro runs on the server, a `console.log()` in Astro frontmatter will output to the **terminal**, not the browser console.

Better/worse? I’m sure you’ll make the right call!

description: Debug in Astro using the Debug component
---

Astro provides several different tools to help you debug your code easier and faster.

## Debugging with `console.log()`

`console.log()` is a simple-but-popular method of debugging your Astro code, especially in the component front matter.  

```astro
---
const sum = (a, b) => a + b;

// Example: Outputs "4" to the CLI terminal
console.log(sum(2, 2));
---
```

A `console.log()` statement in Astro frontmatter will always output to the **terminal** running the Astro CLI. This is because Astro runs on the server, and never in the browser.

> ## Debugging JS Scripts

Reviewer: Should this and the following heading be `<h3>` to nest under “Debugging with `console.log()`”?

```suggestion
### Debugging JS Scripts
```

The headings don’t suggest it but both sections are still focused on `console.log`.

Reviewer 2: I see your point, and can think of 2 or 3 different ways to go on this. I'm going to make a change and see if we like it.  :)

Code that is written or imported inside of an Astro `<script>` tag is run in the browser. Any `console.log()` statements or other debug otuput will be  printed to the console in your browser. 

## Debugging Framework Components

> Framework components (like React and Svelte) are unique: They render server-side by default, meaning that console.log() debug output will be visible in the terminal. However, they can also be hydrated for the browser, which may cause your debug logs to also appear in the browser.

Reviewer: Missing backticks around `console.log`:

```suggestion
Framework components (like React and Svelte) are unique: They render server-side by default, meaning that `console.log()` debug output will be visible in the terminal. However, they can also be hydrated for the browser, which may cause your debug logs to also appear in the browser.
```

This can be useful for debugging differences between the SSR output and the hydrated components in the browser.

## Using the Astro Debug Component

To help you debug your Astro components, Astro provides a built-in [`<Debug />`](/en/reference/builtin-components#debug-) component which renders any value directly into your component HTML template. This is useful for quick debugging in the browser without having to flip back-and-forth between your terminal and your browser.

```astro
---
import { Debug } from 'astro/components';
const sum = (a, b) => a + b;
---

<!-- Example: Outputs "{answer: 4}" to the browser -->
<Debug answer={sum(2, 4)} />
```

> The Debug component also supports object prop syntax for even more flexible and concise debugging:

Reviewer: Will “object prop syntax” be immediately comprehensible? It’s a little jargon-y. Not sure if there’s somewhere we could link to to clarify this or if some rewording would help. Maybe something like

```suggestion
The Debug component also supports shorthand syntax for even more flexible and concise debugging:
```

@sarah11918 You usually manage to rephrase things just right — what do you think?

Reviewer 2: I did something!

```astro
---
import { Debug } from 'astro/components';
const sum = (a, b) => a + b;
const answer = sum(2, 4);
---

> <!-- Example: All three examples are equivilent. -->

Reviewer: Typo:

```suggestion
<!-- Example: All three examples are equivalent. -->
```

<Debug answer={sum(2, 4)} />

> <Debug {answer: sum(2, 4)} />

Reviewer: This example needs double curly braces to work:

```suggestion
<Debug {{ answer: sum(2, 4) }} />
```

Reviewer 2: Checked, and confirmed!
